### PR TITLE
Changed CompoundTag access

### LINKED ISSUE #1118: Item->clearCustomBlockData() fails
### Issue description
https://github.com/pmmp/PocketMine-MP/blob/master/src/pocketmine/item/Item.php#L447

### PR DESCRIPTION
**Pull request by dktapps**
## Introduction
There is a lot of undefinedness around anything that uses NBT in the PocketMine-MP core. This pull request aims to reduce this problem by tightening the restrictions on the way CompoundTag children are accessed, and add methods for accessing them safely instead.

## Relevant issues
- fixed #1118 
- fixed villager profession doesn't make any difference client-side

### API changes
The method of assigning a child tag to a CompoundTag has changed. It is no longer legal to do the following:
- `$nbt->SomeTagName = new IntTag("SomeTagName", 5);`
- `unset($nbt->SomeTagName);`
- `$nbt->SomeTagName->getValue();`

You are instead encouraged to use the new CompoundTag API methods:
- `getTag(string $name)`
- `getCompoundTag(string $name)`
- `setTag(NamedTag $tag)`
- `exists(string $name)`
- `remove(string $name)`

This aims to reduce confusion between CompoundTag offsets and the tags' own names. For example, the following code:
`$nbt->SomeTagName = new IntTag("SomeOtherNameThatDoesntMatch", 5);`
will cause the IntTag to be saved with the name passed in the IntTag constructor, _not_ its field name in the CompoundTag. This duplication/inconsistency caused many bugs throughout the core code.

## Backwards compatibility
This is a backwards-compatibility break due to the API changes noted above.

## Tests
This has only been lightly tested (world loading, enchantments, spawning entities) but there are a lot of changes in this PR and I have no doubt I have introduced problems somewhere. Please test and review carefully.

Areas affected by these changes include:
- Level loading and saving (all formats)
- Entity loading and saving (all)
- Player data loading and saving
- Tile creation, spawning loading and saving.

### CODE REVIEW

**Comment by jasonwynn10** [comment]
I believe this will need another API bump due to the amount of plugin use

**Comment by Sandertv** [comment]
Agreed with @jasonwynn10, this breaks a LOT of plugins, and another API bump should definitely be considered...

**Comment by dktapps** [comment]
well obviously... but you're supposed to use tags, not the master branch. We're due a release very soon anyway.

**Review by matcracker — APPROVED** [review]

File: src/pocketmine/entity/Entity.php [hunk 1]
```diff
@@ -782,34 +783,37 @@ public function saveNBT(){
...
 				$this->setNameTagVisible($this->namedtag["CustomNameVisible"] > 0);
 			}
 		}
 
 		$this->scheduleUpdate();
 
-		if(isset($this->namedtag->ActiveEffects)){
-			foreach($this->namedtag->ActiveEffects->getValue() as $e){
-				$amplifier = Binary::unsignByte($e->Amplifier->getValue()); //0-255 only
+		$effectsTag = $this->namedtag->exists("ActiveEffects");
```
Comment (dktapps) [inline src/pocketmine/entity/Entity.php]: oops

**Pull request closed without merging**